enhancements/cluster-runtime-constraints.md

```diff
@@ -0,0 +1,138 @@
+
+### Goals
+- Provide machanism (including specification and guideline) for cluster admins to specify cluster runtime constraints
+- Provide specification and guideline for author operators to specify runtime contraints/requirements if needed
+- Block the updates/installations of operators based on cluster runtime constraints
+- Allow the cluster runtime constraints to be generic and confugurable by cluster admins
+
+### Non-Goals
+
+- Provide machanism to retrieve/populate cluster runtime constraints automatically
+- Allow cluster runtime constraints to be scoped
+- Allow cluster runtime constraints to come from multiple/different sources
+- Allow optional cluster runetime constraints
+
+
+## Proposal
+
+To provide a generic solution for cluster runtime constaints, a library named [cel-go](https://github.com/google/cel-go) that provides expression evaluation capabilities is utilized in this proposal.
+
+### User Stories
+
+#### Constraints specified by cluster admin
+
+As a cluster admin, I would like to specify a list of cluster runtime constraints that are associated with a given cluster and all installable operators will meet these requirements.
+
+#### Constraint properties specified by operator author
+
+As an operator author, I would like to provide a list of runtime properties that may be needed to be considered before installation to ensure the operator working properly on the cluster.
+
+### Implementation Details
+
+#### Cluster Runtime Constraint Configmap
+
+Cluster admins can specified cluster runtime constraints in a configmap (named `olm-runtime-constraints`) in `olm` namespace using this format:
+
+```yaml=
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: olm-runtime-constraints
+  namespace: olm
+data:
+    properties: <list-of-expression-constraints-in-json>
+```
+
+Each constraint is specified using the following syntax:
+
+```yaml=
+{
+  "type": "olm.constraint",
+  "value": {
+    "evaluator": {
+        "id": "cel",
+    }
+    "source": 'properties.exists(p, p.type == "certified")',
+    "action": {
+        "id": "require"
+    }
+  }
+}
+```
+
+The constraint `type` is `olm.constraint` and the `value` is information associated with constraint. The `value` struct has 3 fields: `evaluator`, `source` and `action`.
+
+The `evaluator` is a struct with `id` field to represent the languague library that will be used to evaluate the expression. At the moment, only `cel-go` library is supported using `cel` as the identifier. More fields are potentially added to expand the supported expression languagues/libraries in the future.
+
+The `source` field is the string of expression that will be evaluated during resolution. Currently, only `cel-go` expressions are supported.
```

Comment:
It's not clear to me the target of the expression.  Is it the [Bundle Properties](https://github.com/operator-framework/enhancements/blob/master/enhancements/properties.md)?  The spec portion of the ClusterServiceVersion?

Reply:
The expression will be evaluated against bundle properties.